### DISCUSSION TITLE
Adding support for route based middleware

### DISCUSSION THREAD

**Discussion by protibimbok** [encode/starlette#1781] [Ideas] [answered]
In other frameworks (i.e. Laravel, Express) there is an way of adding middleware to a route easily which has access to the request object and is able to send response. When I tried this (FastApi -> Starlette) it was missing. These middlewares do make the job a lot easier and in some cases I think it's essential to have.
So I've tried to implement this feature. 

The route middlewares behaves more like an endpoint than a middleware.
These middlewares receive two parameters `request: Request` and `next: async function`

```python
async def check_admin(request: Request, next):
    if request.headers['Authorization'] == 'I am admin':
        request.isAdmin = True
        request.perms = ['a', 'b', 'c']
        return await next()
    else:
        return JSONResponse(['error':'Unauthorized access'])


async def check_perm(request: Request, next):
    if request.isAdmin and 'c' in request.perms:
        return await next()
    else:
        return JSONResponse(['error':'Unauthorized access'])



routes = [
    Route("/admin", endpoint = admin_page, middlewares = check_admin),
    Route(
        "/admin/secret", endpoint = admin_secret,
        middlewares = [check_admin, check_perm] # Middlewares will be ran in the same order
    )
]

```

How's the idea? And can someone help me to run the tests?

**Comment by adriangb** [comment, marked as answer]
Please take a look at https://github.com/encode/starlette/pull/1649

**Reply by protibimbok** [reply]
This is a complete different approach. Sorry previously my code was not formatted. This is pretty close to Express & Laravel and can manipulate the request object.

**Reply by Kludex** [reply]
If you follow the link on the PR description, and do the same on the following... You'll find out what you want...

**Reply by protibimbok** [reply]
@Kludex Sorry I could not understand. Did you mean this is already implemented?

**Reply by Kludex** [reply]
No, I mean that you didn't follow the links. There's a history behind the PR Adrian sent.

See https://github.com/encode/starlette/pull/1286

**Reply by protibimbok** [reply]
Thank you. I have read these history. That does work for me, but can you please take a look at this [Pull Request](https://github.com/encode/starlette/pull/1789)? It's a different approach and I think it's much cleaner...

**Comment by protibimbok** [comment]
Test fails for trio when ran together, but succeeds when ran separately!
![image](https://user-images.githubusercontent.com/33311165/182759421-30db0346-25e7-4bff-8714-8f7ea3d4b491.png)
![image](https://user-images.githubusercontent.com/33311165/182759454-8b8b32d8-51e8-47d7-ba60-0b6ebaba4e1d.png)
